cpp/src/arrow/util/thread_pool.h

```diff
@@ -104,15 +104,22 @@ class ARROW_EXPORT Executor {
   template <typename T>
   Future<T> Transfer(Future<T> future) {
     auto transferred = Future<T>::Make();
-    future.AddCallback([this, transferred](const Result<T>& result) mutable {
+    auto callback = [this, transferred](const Result<T>& result) mutable {
       auto spawn_status = Spawn([transferred, result]() mutable {
         transferred.MarkFinished(std::move(result));
       });
       if (!spawn_status.ok()) {
         transferred.MarkFinished(spawn_status);
       }
-    });
-    return transferred;
+    };
+    auto callback_factory = [&callback]() { return callback; };
+    if (future.TryAddCallback(callback_factory)) {
+      return transferred;
+    }
+    // If the future is already finished and we aren't going to force spawn a thread
+    // then we don't need to add another layer of callback and can return the original
+    // future
```

Comment:
In this case the future is already finished.  There is nothing we can do.  The old implementation suffered from the same problem.  `Transfer(finished_fut).AddCallback(...)` will run the callback synchronously on the calling thread regardless of what you do inside of `Transfer(finished_fut)`.

So far this isn't a problem, the only place we call transfer is from the CPU thread pool to take something off the I/O thread pool.  If the future is finished then the callback will run synchronously on the CPU thread pool anyways.  The only place it could really be an issue I suppose is if you were trying to transfer it onto a different thread pool than the calling thread pool.

That being said I could revert all the changes to this method.  They were leftover from an earlier misunderstanding and they don't change the behavior of `Transfer` at all.  I simply left it this way because I felt it was clearer about the effects.

Reply:
> `Transfer(finished_fut).AddCallback(...)` will run the callback synchronously on the calling thread 

Are you sure it does? `transferred.MarkFinished` is called on the thread pool, so it would depend on whether it executes because the caller calls `AddCallback`. In a way, it's even more underterministic than the current version, though, so arguably worse.

That said, I do think it's a problem that this doesn't reliably transfer the callbacks to the thread pool. Perhaps this isn't the right API after all: instead of `ThreadPool::Transfer(future)`, do we instead want  `Future::AddCallback(Executor*, Callback&&)` and/or `Future::Then(Executor*, Callback&&)`?

Reply:
I agree with the new API.  C# has that API as well (Task::ContinueWith can take a scheduler) so there is some precedence.  I'll add a follow-up PR.  I don't believe it is necessary for the current work though.

You are correct, the old method could possible introduce a new thread task spawned in some situations where this doesn't (e.g. if the callback is added very quickly after calling transfer before the TP has a chance to schedule the "mark finished" task).  I don't really see any reason this is desirable as it can only mean more thread pool tasks without any benefit.  So I will leave this as is for now.

Reply:
>  I'll add a follow-up PR. I don't believe it is necessary for the current work though.

Sounds good.

Reply:
ARROW-12091